# Telecom X — Customer Churn Prediction

Binary classification: predict which customers cancel. Logistic Regression vs. Random Forest, evaluated on a held-out test set.

Data: `telecomx_datos_tratados.csv` (7,043 labeled customers), produced by the [EDA notebook](https://github.com/MarlonPC32/telecomx-churn-analysis). 224 records with unknown churn labels were excluded there — they are not in this file.

In [1]:
import pandas as pd

df = pd.read_csv("telecomx_datos_tratados.csv")
print(df.shape)
print(df["Churn"].value_counts(normalize=True).round(4))
df.head()

(7043, 22)
Churn
0    0.7346
1    0.2654
Name: proportion, dtype: float64


,customerID,Churn,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,Charges.Monthly,Charges.Total,Cuentas_Diarias
0,0002-ORFBO,0,Female,0,Yes,Yes,9,Yes,No,dsl,...,No,Yes,Yes,No,one year,Yes,mailed check,65.6,593.30,2.186667
1,0003-MKNFE,0,Male,0,No,No,9,Yes,Yes,dsl,...,No,No,No,Yes,month-to-month,No,mailed check,59.9,542.40,1.996667
2,0004-TLHLJ,1,Male,0,No,No,4,Yes,No,fiber optic,...,Yes,No,No,No,month-to-month,Yes,electronic check,73.9,280.85,2.463333
3,0011-IGKFF,1,Male,1,Yes,No,13,Yes,No,fiber optic,...,Yes,No,Yes,Yes,month-to-month,Yes,electronic check,98.0,1237.85,3.266667
4,0013-EXCHZ,1,Female,1,Yes,No,3,Yes,No,fiber optic,...,No,Yes,Yes,No,month-to-month,Yes,mailed check,83.9,267.40,2.796667


## Feature selection

- Drop `customerID`: an identifier with no predictive signal.
- Drop `Cuentas_Diarias`: it is exactly `Charges.Monthly / 30` (correlation 1.00 with the monthly column). A pure rescale adds no information to either model — in the linear model it is exactly redundant, and in the tree model it only fragments feature-importance scores across two identical signals.

In [2]:
print("corr(Cuentas_Diarias, Charges.Monthly) =",
      round(df["Cuentas_Diarias"].corr(df["Charges.Monthly"]), 4))

df = df.drop(columns=["customerID", "Cuentas_Diarias"])
X = pd.get_dummies(df.drop(columns=["Churn"]), drop_first=True)
y = df["Churn"]
print("predictors:", X.shape[1], "| rows:", len(y))

corr(Cuentas_Diarias, Charges.Monthly) = 1.0
predictors: 30 | rows: 7043


## Train/test split (stratified)

Stratification keeps the 26.5% churn rate identical in train and test — without it, the split can drift several points on an imbalanced target.

In [3]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.30, random_state=42, stratify=y)
print("train churn rate:", round(y_train.mean(), 4),
      "| test churn rate:", round(y_test.mean(), 4),
      "| test n:", len(y_test))

train churn rate: 0.2653 | test churn rate: 0.2655 | test n: 2113


## Baseline: majority-class classifier

If we predicted "retained" for everyone, accuracy would equal the majority share. Any model must beat this to demonstrate it learned anything.

In [4]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

baseline = DummyClassifier(strategy="most_frequent")
baseline.fit(X_train, y_train)
base_acc = accuracy_score(y_test, baseline.predict(X_test))
print(f"majority-class baseline accuracy: {base_acc:.4f}")

majority-class baseline accuracy: 0.7345


## Models: scaled pipelines

Numeric features span very different ranges (tenure 0–72, charges in the thousands). An unscaled logistic regression hits the iteration limit without converging, so scaling goes **inside a pipeline** — fit on training data only, so no information leaks into the test set.

In [5]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

logreg = make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000))
rf = make_pipeline(StandardScaler(), RandomForestClassifier(random_state=42))

logreg.fit(X_train, y_train)
rf.fit(X_train, y_train)

# convergence check: did the optimizer finish before the iteration cap?
n_iter = logreg.named_steps["logisticregression"].n_iter_.max()
print("logreg iterations used:", n_iter, "| converged:", n_iter < 1000)

logreg iterations used: 36 | converged: True


In [6]:
from sklearn.metrics import (accuracy_score, precision_score, recall_score,
                                   f1_score, confusion_matrix, classification_report)

def report(name, model):
    p = model.predict(X_test)
    print(f"--- {name} ---")
    print(f"Accuracy:  {accuracy_score(y_test, p):.4f}")
    print(f"Precision: {precision_score(y_test, p):.4f}  (of predicted churners, how many actually churned)")
    print(f"Recall:    {recall_score(y_test, p):.4f}  (of actual churners, how many were caught)")
    print(f"F1:        {f1_score(y_test, p):.4f}")
    print("Confusion matrix [tn, fp / fn, tp]:")
    print(confusion_matrix(y_test, p))
    print()

report("Logistic Regression", logreg)
report("Random Forest", rf)
print(f"Majority-class baseline accuracy: {base_acc:.4f}")

--- Logistic Regression ---
Accuracy:  0.7979
Precision: 0.6402  (of predicted churners, how many actually churned)
Recall:    0.5455  (of actual churners, how many were caught)
F1:        0.5890
Confusion matrix [tn, fp / fn, tp]:
[[1380  172]
 [ 255  306]]

--- Random Forest ---
Accuracy:  0.7842
Precision: 0.6218  (of predicted churners, how many actually churned)
Recall:    0.4777  (of actual churners, how many were caught)
F1:        0.5403
Confusion matrix [tn, fp / fn, tp]:
[[1389  163]
 [ 293  268]]

Majority-class baseline accuracy: 0.7345


In [7]:
print(classification_report(y_test, logreg.predict(X_test), digits=3))
print(classification_report(y_test, rf.predict(X_test), digits=3))

              precision    recall  f1-score   support

           0      0.844     0.889     0.866      1552
           1      0.640     0.545     0.589       561

    accuracy                          0.798      2113
   macro avg      0.742     0.717     0.728      2113
weighted avg      0.790     0.798     0.792      2113



              precision    recall  f1-score   support

           0      0.826     0.895     0.859      1552
           1      0.622     0.478     0.540       561

    accuracy                          0.784      2113
   macro avg      0.724     0.686     0.700      2113
weighted avg      0.772     0.784     0.774      2113



## What the models associate with churn

Two honest caveats before reading these tables:

1. Random Forest impurity importance favors high-cardinality features and splits correlated predictors' credit — treat it as a rough ranking, not a measurement.
2. Logistic coefficients are on standardized features, so magnitudes are comparable — but correlated predictors (tenure vs. charges) split credit, and these are associations, not causal effects. Shown below are the largest positive and negative associations, not a causal ranking.

In [8]:
import pandas as pd

importances = pd.Series(rf.named_steps["randomforestclassifier"].feature_importances_,
                        index=X.columns).sort_values(ascending=False)
print("Random Forest — top 10 by impurity importance:")
print(importances.head(10).round(4))

Random Forest — top 10 by impurity importance:
Charges.Total                     0.1906
tenure                            0.1730
Charges.Monthly                   0.1631
PaymentMethod_electronic check    0.0380
InternetService_fiber optic       0.0368
Contract_two year                 0.0300
gender_Male                       0.0282
OnlineSecurity_Yes                0.0266
PaperlessBilling_Yes              0.0264
TechSupport_Yes                   0.0249
dtype: float64


In [9]:
coefs = pd.Series(logreg.named_steps["logisticregression"].coef_[0], index=X.columns)
print("Logistic Regression — strongest positive associations:")
print(coefs.sort_values(ascending=False).head(8).round(4))
print()
print("Logistic Regression — strongest negative (protective) associations:")
print(coefs.sort_values().head(8).round(4))

Logistic Regression — strongest positive associations:
Charges.Total                     0.7417
InternetService_fiber optic       0.5474
PaperlessBilling_Yes              0.2020
PaymentMethod_electronic check    0.1725
StreamingTV_Yes                   0.1660
MultipleLines_Yes                 0.1628
StreamingMovies_Yes               0.1460
SeniorCitizen                     0.0627
dtype: float64

Logistic Regression — strongest negative (protective) associations:
tenure                             -1.4917
Contract_two year                  -0.5624
Charges.Monthly                    -0.4698
Contract_one year                  -0.3014
TechSupport_Yes                    -0.1607
OnlineSecurity_Yes                 -0.1080
Dependents_Yes                     -0.0935
OnlineBackup_No internet service   -0.0688
dtype: float64


## Findings and limitations

- Both models beat the majority-class baseline on accuracy, but the useful comparison is on the churn class: logistic regression catches more actual churners (higher recall) at a similar precision cost.
- Spending behavior (total/monthly charges) and tenure rank highest in the tree model. In the logistic model, the largest absolute standardized coefficients are tenure (-1.49, protective) and total charges (+0.74); among service features, fiber-optic internet, paperless billing, and electronic check carry the largest positive associations.
- The two models disagree on exact rankings — expected, since they measure different things (split utility vs. linear association in scaled units).
- Limitations: single 70/30 split (no cross-validation); default 0.5 decision threshold (not tuned to any business cost of missed churners vs. unnecessary outreach); no defined scoring date or churn horizon, so this classifies a snapshot rather than operating as a validated early-warning system.
- Built for the Telecom X Data Science Challenge (Oracle Next Education).